go/osagent/osagent.go

```diff
@@ -645,3 +745,609 @@ func AbortSeed(seedId string) error {
+	var cmd string
+	BackupFolder = path.Join(config.Config.MySQLBackupDir, time.Now().Format("20060102-150405"))
+	if err := os.Mkdir(BackupFolder, 0755); err != nil {
+		return "", log.Errore(err)
+	}
+	if !contains(seedMethod, seedMethods) {
+		return "", log.Errorf("Unsupported seed method")
+	}
+	if databases != "" {
+		availiableDatabases, _ := dbagent.GetMySQLDatabases()
+		for _, db := range strings.Split(databases, ",") {
+			if !contains(db, availiableDatabases) {
+				return "", log.Errorf("Cannot backup database %+v. Database doesn't exists", db)
+			}
+		}
+		//if we don't already have mysql database in databases, add it
+		if !strings.Contains(databases, "mysql") {
+			databases += ",mysql"
+		}
+		//and if we are on 5.7 we need to add sys db because sometimes during mysql_upgrade run we can get errors like mysql_upgrade: [ERROR] 1813: Tablespace '`sys`.`sys_config`' exists
+		version, _ := dbagent.GetMySQLVersion()
+		if version == "5.7" && !strings.Contains(databases, "sys") {
+			databases += ",sys"
+		}
+	}
+	if err := dbagent.ManageReplicationUser(); err != nil {
+		return "", log.Errore(err)
+	}
+	switch seedMethod {
+	case "xtrabackup":
+		cmd = fmt.Sprintf("xtrabackup --backup --user=%s --password=%s --port=%d --parallel=%d --target-dir=%s --databases='%s'",
+			config.Config.MySQLTopologyUser, config.Config.MySQLTopologyPassword, config.Config.MySQLPort, config.Config.XtrabackupParallelThreads, BackupFolder, strings.Replace(databases, ",", " ", -1))
+	case "mydumper":
+		cmd = fmt.Sprintf("mydumper --user=%s --password=%s --port=%d --threads=%d --outputdir=%s --triggers --events --routines --regex='(%s)'",
+			config.Config.MySQLTopologyUser, config.Config.MySQLTopologyPassword, config.Config.MySQLPort, config.Config.MyDumperParallelThreads, BackupFolder, strings.Replace(databases, ",", "\\.|", -1)+"\\.")
+		if config.Config.CompressLogicalBackup {
+			cmd += fmt.Sprintf(" --compress")
+		}
+		if config.Config.MyDumperRowsChunkSize != 0 {
+			cmd += fmt.Sprintf(" --rows=%d", config.Config.MyDumperRowsChunkSize)
+		}
+	case "mysqldump":
+		if databases == "" {
+			cmd = fmt.Sprintf("mysqldump --user=%s --password=%s --port=%d --single-transaction --master-data=1 --routines --events --triggers --all-databases",
+				config.Config.MySQLTopologyUser, config.Config.MySQLTopologyPassword, config.Config.MySQLPort)
+		} else {
+			cmd = fmt.Sprintf("mysqldump --user=%s --password=%s --port=%d --single-transaction --master-data=1 --routines --events --triggers --databases %s",
+				config.Config.MySQLTopologyUser, config.Config.MySQLTopologyPassword, config.Config.MySQLPort, strings.Replace(databases, ",", " ", -1))
+		}
+		if config.Config.CompressLogicalBackup {
+			cmd += fmt.Sprintf(" | gzip > %s", path.Join(BackupFolder, mysqlbackupCompressedFileName))
+		} else {
+			cmd += fmt.Sprintf(" > %s", path.Join(BackupFolder, mysqlbackupFileName))
+		}
+	}
```

Comment:
Let's extract the command creation to a set of functions. This function is already too big with too much text formatting.

Reply:
fixed in 3ce673a